keps/sig-node/20200619-provisioning-resources-with-numa-topology.md

```diff
@@ -0,0 +1,240 @@
+To avoid such kind of problem scheduler should choose a node considering topology
+of the node and TopologyManager policy on the worker node.
+
+## Motivation
+
+For the sake of scheduling which is topology aware, resources with topology
+information should be provisioned.
+This KEP describes how it would be implemented.
+
+### Goals
+
+Provisioning resources with topology information.
+
+### Non-Goals
+
+ - modification of any public API
+ - improving and as a result modification of the TopologyManager and its policies
+
+## Proposal
+
+Introduce standalone daemon which will be run on each node in the cluster as daemonset. It should
+collect resources for running pods and provide it with CRD.
+
+## Design Details
+
+The design consists of part which describes how datum collected
+and how it was provided.
+
+Resources used by the pod could be obtained by podresources interface
+of the kubelet and by CRI from container runtime.
+
+To calculate available resources need to know all resources
+which could be used by kubernetes. It could be calculated by
+substracting resources of kube cgroup and system cgroup from
+system allocatable resources.
+
+### Design based on CRI
+
+The containerStatusResponse returned as a response to the ContainerStatus rpc contains `Info` field which is used by the container runtime for capturing ContainerInfo.
+```go
+message ContainerStatusResponse {
+      ContainerStatus status = 1;
+      map<string, string> info = 2;
+}
+```
+
+Containerd has been used as the container runtime in the initial investigation. The internal container object info
+[here](https://github.com/containerd/cri/blob/master/pkg/server/container_status.go#L130)
+
+The Daemon set is responsible for the following:
+
+- Parsing the info field to obtain container resource information
+- Identifying NUMA nodes of the allocated resources
+- Identifying total number of resources allocated on a NUMA node basis
+- Detecting Node resource capacity on a NUMA node basis
+- Updating the CRD instance per node indicating available resources on NUMA nodes, which is referred to the scheduler
+
+
+#### Drawbacks
+
+The content of the `info` field is free form, unregulated by the API contract. So, CRI-compliant container runtime engines are not required to add any configuration-specific information, like for example cpu allocation, here. In case of containerd container runtime, the Linux Container Configuration is added in the `info` map depending on the verbosity setting of the container runtime engine.
+
+There is currently work going on in the community as part of the the Vertical Pod Autoscaling feature to update the ContainerStatus field to report back containerResources
+[KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md).
+
+
+### Design based on podresources interface of the kubelet
+
+Podresources interface of kubelet described in
```

Comment:
i think @dashpole should review this carefully.

Reply:
yes, we already started a KEP (https://github.com/kubernetes/enhancements/pull/1884). It covers our requirements (maybe except memory).